tests/rig/tests/test_runtime.py

```
@@ -62,3 +62,48 @@ def test_platform_endpoints_from_env_uses_defaults(monkeypatch) -> None:
+
+
+def test_login_provider_resolves_dotted_path(monkeypatch) -> None:
+    monkeypatch.setenv("UNSTRACT_E2E_LOGIN_PROVIDER", "json.loads")
+    import json
+
+    assert _load_login_provider() is json.loads
+
+
+def test_login_provider_rejects_bare_name(monkeypatch) -> None:
+    monkeypatch.setenv("UNSTRACT_E2E_LOGIN_PROVIDER", "loads")
+    with pytest.raises(ValueError, match="module.func"):
```

Ruff (0.16.0)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)